### PR TITLE
Bazel initial commit with updates

### PR DESCRIPTION
**Pull request by mobileink**
All file additions, no changes to existing code.  The vast majority of the new files are autogenerated in `bzl/cargo/remote`.

NOTE: some of the packages not needed by Mina do not build.  To excluded them their directories are listed in `.bazelignore`.

### CODE REVIEW

File: WORKSPACE.bazel [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+
+load("@io_bazel_rules_rust//:workspace.bzl", "bazel_version")
+bazel_version(name = "bazel_version")
+
+################################################################
+################################################################
+# For local (independent) builds, this local repo allows us to import
+# marlin, which imports zexe. Some of marlin's targets depend on
+# @zexe// packages; this repo allows Bazel to resolve them here. In
+# addition, zexe targets that depend on both zexe and marlin packages
```
Comment (mrmr1993) [inline WORKSPACE.bazel]: This is no longer true now `snarky-bn382` is deprecated. Can we remove the dependency-cycle-break stuff for marlin/zexe here?

File: algebra/BUILD.bazel [hunk 2]
```diff
@@ -0,0 +1,89 @@
+load(
+    "@io_bazel_rules_rust//rust:rust.bzl",
+    "rust_library",
+)
+
+rust_library(
+    name = "algebra-snarky",
```
Comment (mrmr1993) [inline algebra/BUILD.bazel]: I am deeply, deeply unhappy about this. Please tear it out and specify the features where the library is required with those features.

**Review by mrmr1993 — CHANGES_REQUESTED** [review]

File: algebra/BUILD.bazel (hunk 2 above)
Comment (mobileink) [reply algebra/BUILD.bazel]: done

File: WORKSPACE.bazel (hunk 1 above)
Comment (mobileink) [reply WORKSPACE.bazel]: done

File: cstubs_applicative/cstubs_applicative.ml [hunk 3]
```diff
@@ -0,0 +1,113 @@
+open Base
```
Comment (mrmr1993) [inline cstubs_applicative/cstubs_applicative.ml]: This is a second copy of a library we're not using at all. Please delete it

File: bzl/external/rust_deps.bzl [hunk 4]
```diff
@@ -0,0 +1,24 @@
+load("@bazel_tools//tools/build_defs/repo:git.bzl", "git_repository")  # buildifier: disable=load
+load("@bazel_tools//tools/build_defs/repo:http.bzl", "http_archive")  # buildifier: disable=load
+load("@bazel_tools//tools/build_defs/repo:utils.bzl", "maybe")  # buildifier: disable=load
+
+def rust_bootstrap():
+    """This bootstraps (loads) Golang rules and repos"""
```
Comment (mrmr1993) [inline bzl/external/rust_deps.bzl]: Nit: `s/Golang/Rust`

File: bzl/external/rust_deps.bzl [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+        name = "io_bazel_rules_rust",
+        sha256 = "618cba29165b7a893960de7bc48510b0fb182b21a4286e1d3dbacfef89ace906",
+        strip_prefix = "rules_rust-5998baf9016eca24fafbad60e15f4125dd1c5f46",
+        urls = [
+            # Master branch as of 2020-09-24
+            "https://github.com/bazelbuild/rules_rust/archive/5998baf9016eca24fafbad60e15f4125dd1c5f46.tar.gz",
+        ],
+    )
+
+    maybe(
```
Comment (mrmr1993) [inline bzl/external/rust_deps.bzl]: Nothing that we build in this repo now depends on marlin. Delete this.

File: WORKSPACE.bazel [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+        "https://mirror.bazel.build/github.com/bazelbuild/bazel-skylib/releases/download/1.0.3/bazel-skylib-1.0.3.tar.gz",
+    ],
+    sha256 = "1c531376ac7e5a180e0237938a2536de0c54d93f5c278634818e0efc952dd56c",
+)
+
+load("@bazel_skylib//:workspace.bzl", "bazel_skylib_workspace")
+bazel_skylib_workspace()
+
+############### Bootstrap OPAM Rules ###############
+git_repository(
```
Comment (mrmr1993) [inline WORKSPACE.bazel]: We don't build any ocaml code from this repo now, this can go.

File: WORKSPACE.bazel [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+    name = "obazl_rules_opam",
+    remote = "https://github.com/obazl/rules_opam",
+    branch = "main",
+)
+load("@obazl_rules_opam//opam:bootstrap.bzl", "opam_configure")
+load("//bzl:opam.bzl", "opam")
+opam_configure(opam = opam)
+
+############### Bootstrap OCaml Rules ###############
```
Comment (mrmr1993) [inline WORKSPACE.bazel]: This too

File: snarky-bn382/caml/BUILD.bazel [hunk 8]
```diff
@@ -0,0 +1,28 @@
+load(
```
Comment (mrmr1993) [inline snarky-bn382/caml/BUILD.bazel]: This library is deprecated, unused. Delete it.

File: bzl/external/rust_deps.bzl (hunk 5 above)
Comment (mrmr1993) [reply bzl/external/rust_deps.bzl]: @mobileink This is the last thing that really needs to go before we merge, everything else can be done later

File: cstubs_applicative/cstubs_applicative.ml (hunk 3 above)
Comment (mobileink) [reply cstubs_applicative/cstubs_applicative.ml]: Done

File: bzl/external/rust_deps.bzl (hunk 4 above)
Comment (mobileink) [reply bzl/external/rust_deps.bzl]: Done

File: bzl/external/rust_deps.bzl (hunk 5 above)
Comment (mobileink) [reply bzl/external/rust_deps.bzl]: Done

File: WORKSPACE.bazel (hunk 7 above)
Comment (mobileink) [reply WORKSPACE.bazel]: Done

**Review by mrmr1993 — APPROVED** [review]

**Pull request merged**